### PR TITLE
[PM-1028] Simple kademlia refresh process

### PR DESCRIPTION
**Pull request by KonradStaniec**
Implements kademlia refresh process in simple version used in other p2p implementation i.e performing lookup for random targets:
https://github.com/ethereum/go-ethereum/blob/fa538ee7ed04b1a5e101938d64805d3fcd0eb697/p2p/discover/table.go#L298
https://github.com/status-im/nim-eth/blob/e6164996c8e1ce1a33c17250935fc985b9120830/eth/p2p/discovery.nim#L284

### CODE REVIEW

File: scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala [hunk 1]
```diff
@@ -9,7 +10,11 @@ import scodec.bits.BitVector
   *
```
Comment (jtownson) [inline scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala:9]: The comment that this is a 'skeletal' kbuckets implementation is now wrong.

File: scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala [hunk 2]
```diff
@@ -20,11 +21,33 @@ class KRouter[A](
...
+    */
+  private def startRefreshCycle(): Task[Unit] = {
+    Observable
+      .intervalWithFixedDelay(config.refreshRate, config.refreshRate)
+      .consumeWith(Consumer.foreachTask { _ =>
+        lookup(KBuckets.generateRandomId(config.nodeRecord.id.length, rnd)).map(_ => ())
+      })
+  }
+
   // TODO[PM-1035]: parallelism should be configured by library user
```
Comment (jtownson) [inline scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:51]: Probably it should be some function of alpha, which is already user provided.

File: scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala [hunk 3]
```diff
@@ -386,14 +409,20 @@ class KRouter[A](
...
 object KRouter {
+  import scala.concurrent.duration._
   case class Config[A](
       nodeRecord: NodeRecord[A],
       knownPeers: Set[NodeRecord[A]],
       alpha: Int = 3,
       k: Int = 20,
-      serverBufferSize: Int = 2000
+      serverBufferSize: Int = 2000,
+      refreshRate: FiniteDuration = 15.minutes
```
Comment (jtownson) [inline scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:429]: This config is somehow a bit leaky. Suggest we remove the buffer size parameter and try to understand the kademlia _m_ parameter. To date, we support kademlia's k parameter (measuring the probability of a key being found given some assumptions about node stability) and alpha (concurrency parameter).

IMO, the parameter we might want to add now is a node _lifetime_ parameter. I.e. some measure of the time that a node can appear in lookup results _after_ that node has been switched off. I _think_ the original kademlia implementation has a parameter called _m_ (measured in minutes). It seems related, unfortunately my knowledge here is hazy.

In the context of this PR, if the only lookups are those from your refresh process then the expiry time of a key would to be dictated by the frequency of the refresh process. For future PRs, other refresh possibilities exist such as a node republishing its ID to the network. I think we would probably want the same lifetime parameter to apply in both cases (since both are just implementation strategies for dropping expired keys).

TLDR; consider using a more generic parameter name, possibly _m_.

**Review by jtownson — APPROVED** [review]
Looks good.

File: scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala (hunk 2 above)
Comment (KonradStaniec) [reply scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:51]: it could be a good idea if will leave some note about this under the task `PM-1035`

File: scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala (hunk 3 above)
Comment (KonradStaniec) [reply scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:429]: tbh for now i would leave it as `refreshRate` because: 
- i could not find proper description of `m` in original paper
- this name  (or similiar ) is used in other kademlia implementations which gives some familiarity to the user
- original paper mentions only using lookups as refresh process. I do not have nothing against other options like ` node republishing its ID to the network`  but i would rather rename this parameter when/if we will start considering those in future
